shared/src/types/ethereum_events.rs

```diff
@@ -1,214 +1,321 @@
+    UpgradedContract {
+        /// Name of the contract
+        #[allow(dead_code)]
+        name: String,
+        /// Address of the contract on Ethereum
+        #[allow(dead_code)]
+        address: EthAddress,
+    },
+    /// Event indication a new Ethereum based token has been whitelisted for
+    /// transfer across the bridge
+    UpdateBridgeWhitelist {
+        /// Monotonically increasing nonce
+        #[allow(dead_code)]
+        nonce: Uint,
+        /// Tokens to be allowed to be transferred across the bridge
+        #[allow(dead_code)]
+        whitelist: Vec<TokenWhitelist>,
+    },
 }
 
 impl EthereumEvent {
     /// SHA256 of the Borsh serialization of the [`EthereumEvent`].
+    #[allow(dead_code)]
     fn hash(&self) -> Result<Hash, std::io::Error> {
         let bytes = self.try_to_vec()?;
         Ok(Hash::sha256(&bytes))
     }
 }
 
-/// Representation of address on Ethereum
+/// An event transferring some kind of value from Ethereum to Anoma
 #[derive(
-    Clone, PartialEq, Eq, Debug, BorshSerialize, BorshDeserialize, BorshSchema,
+    Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, BorshSchema,
 )]
-pub struct EthAddress(pub [u8; 20]);
+pub struct TransferToNamada {
+    /// Quantity of the ERC20 token in the transfer
+    pub amount: Amount,
+    /// Address of the smart contract issuing the token
+    pub asset: EthAddress,
+    /// The address receiving wrapped assets on Anoma
+    pub receiver: Address,
+}
 
 /// An event transferring some kind of value from Ethereum to Anoma
 #[derive(
-    Debug, PartialEq, Eq, Clone, BorshSerialize, BorshDeserialize, BorshSchema,
+    Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, BorshSchema,
 )]
-pub struct TransferToNamada {
-    /// Quantity of ether in the transfer
+pub struct TransferToEthereum {
+    /// Quantity of wrapped Asset in the transfer
     pub amount: Amount,
-    /// Address on Ethereum of the asset
-    pub asset: EthereumAsset,
-    /// The Namada address receiving wrapped assets on Anoma
-    pub receiver: Address,
+    /// Address of the smart contract issuing the token
+    pub asset: EthAddress,
+    /// The address receiving assets on Ethereum
+    pub receiver: EthAddress,
 }
 
-/// Represents Ethereum assets on the Ethereum blockchain
+/// struct for whitelisting a token from Ethereum.
+/// Includes the address of issuing contract and
+/// a cap on the max amount of this token allowed to be
+/// held by the bridge.
 #[derive(
-    Debug, PartialEq, Eq, Clone, BorshSerialize, BorshDeserialize, BorshSchema,
+    Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, BorshSchema,
 )]
-pub enum EthereumAsset {
-    /// An ERC20 token and the address of its contract
-    ERC20(EthAddress),
+#[allow(dead_code)]
+pub struct TokenWhitelist {
+    /// Address of Ethereum smart contract issuing token
+    pub token: EthAddress,
+    /// Maximum amount of token allowed on the bridge
+    pub cap: Amount,
```

Comment:
I think we should probably stick with the Amounts type unless there is a compelling reason not to. It is the default type throughout the code base for any token amount. Also, micros just means we use 6 (or maybe it's 5) digits of decimal precision.

Reply:
`Amount` works functionally but the term `micro` doesn't make full sense in the context of ERC20 imo. We should have a struct something like

```rust
pub struct Amount {
    units: u64,
}
```

and deal only with whatever the smallest unit of the ERC20 is (since I don't think we should ever be dealing with smaller than that).

I need to come back to this as I'm not familiar enough with how `Amount` works in the context of our tokens itself right now, or ERC20 supply units themselves.

Reply:
Maybe relevant (or might not be), ERC20 tokens can have custom "decimals", the most common being 18. This is more of a UX thing though and shouldn't affect our model of what an amount is, without thinking about it more.